crates/bevy_ecs/src/query/state.rs

```diff
@@ -1276,43 +1339,156 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
             >())),
         }
     }
+
+    /// Gets the reasons an [`Entity`] might not match this query, see [`QueryEntityMismatchDetail`]
+    pub fn get_mismatches_detail(
+        &self,
+        components: &Components,
+        entity_location: EntityLocation,
+        archetype: &Archetype,
+    ) -> Vec<QueryEntityComponentMismatch> {
+        let mut mismatch_details = Vec::new();
+
+        for components_access in &self.component_access.filter_sets {
+            let query_with_components = components_access.with.clone();
+            let query_without_components = components_access.without.clone();
+            let mut query_unmatched_components = query_with_components.clone();
+
+            let mut query_without_components_in_entity = Vec::new();
+            for component_id in archetype.components() {
+                if query_without_components.contains(component_id.index()) {
+                    query_without_components_in_entity.push((
+                        components.get_info(component_id).unwrap().name().to_owned(),
+                        component_id,
+                    ));
+                } else if query_with_components.contains(component_id.index()) {
+                    query_unmatched_components.set(component_id.index(), false);
+                }
+            }
+
+            let query_with_components_not_in_entity = query_unmatched_components
+                .ones()
+                .map(|component_id| {
+                    let component_id = ComponentId::new(component_id);
+                    (
+                        components.get_info(component_id).unwrap().name().to_owned(),
+                        component_id,
+                    )
+                })
+                .collect();
+
+            mismatch_details.push(QueryEntityComponentMismatch {
+                query_with_components_not_in_entity,
+                query_without_components_in_entity,
+                entity_archetype: entity_location.archetype_id,
+                query_with_components,
+                query_without_components,
+            });
+        }
+
+        mismatch_details
+    }
 }
 
 /// An error that occurs when retrieving a specific [`Entity`]'s query result from [`Query`](crate::system::Query) or [`QueryState`].
-// TODO: return the type_name as part of this error
-#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum QueryEntityError {
```

Comment:
Same remark as for `QueryComponentError`. This type should be converted to a `struct`, instead of having every variants hold a `QueryEntityErrorDetail`.